File: android/core/domain/src/main/kotlin/com/cogra/domain/signing/RegistrationSigner.kt
// One pass over the applicant flow (auth.md "Application", "Approval
// and landing"): read the me-driven status, repair whatever this device
// can repair — a minted-but-unattached key, a parked backup blob — and
// sign the staged Registration the moment it appears, through the same
// session-authorized legs as any other write. Every step tolerates a
// lost response: `advance()` is safe to call repeatedly and continues
// from wherever the flow stopped.

package com.cogra.domain.signing

import com.cogra.domain.AccountState
import com.cogra.domain.Outcome
import com.cogra.domain.UserError
import com.cogra.domain.identity.KeyCeremony
import com.cogra.domain.repo.OnboardingRepository
import com.cogra.domain.store.IdentityStore
import java.time.Instant
import javax.inject.Inject

/** Where the applicant flow stands after one `advance()` pass. */
sealed interface RegistrationProgress {
    /**
     * Unapproved, with the two approvability proofs as the server sees
     * them; [keyOnDevice] distinguishes "run the ceremony here" from
     * "the key lives on another device" — a slot key that does not
     * match the account's attached key counts as not-on-device.
     */
    data class AwaitingApproval(
        val emailVerified: Boolean,
        val keyAttached: Boolean,
        val keyOnDevice: Boolean,
    ) : RegistrationProgress

    /** Approved and signed (or mid-seal); landing is the server's move. */
    data object AwaitingLanding : RegistrationProgress

    /** The staged Registration awaits a signature this device cannot make. */
    data object AwaitingSigningKey : RegistrationProgress

    /** No live application — expired unapproved, or reaped. A fresh invite re-arms. */
    data object NeedsInvite : RegistrationProgress

    /** Landed: the account is a member; onboarding is over. */
    data object Member : RegistrationProgress

    /** The device refused to sign the sealed Registration. */
    data class RejectedByDevice(val reason: String) : RegistrationProgress

    /** The API refused a step. */
    data class Refused(val errors: List<UserError>) : RegistrationProgress

    /** Transport failed; call `advance()` again. */
    data class Failed(val cause: Exception) : RegistrationProgress
}

class RegistrationSigner @Inject constructor(
    private val onboarding: OnboardingRepository,
    private val writeSigner: WriteSigner,
    private val identity: IdentityStore,
    private val ceremony: KeyCeremony,
) {
    /**
     * One pass over the flow: poll, repair, sign whatever awaits a
     * signature. Idempotent from the caller's view — the next pass
     * continues wherever this one stopped.
     */
    suspend fun advance(): RegistrationProgress {
        val status = when (val read = onboarding.applicationStatus()) {
            is Outcome.Success -> read.value
            is Outcome.Refused -> return RegistrationProgress.Refused(read.errors)
            is Outcome.Failed -> return RegistrationProgress.Failed(read.cause)
        }
        // A session exists from registration on, so a parked blob never
        // waits on a milestone — flush it on every pass until it lands.
        ceremony.uploadPendingBackup()
        if (status.accountState == AccountState.MEMBER) return RegistrationProgress.Member
        val staged = status.stagedRegistration
        if (staged != null) {
            // The same predicate as keyOnDevice: signing with a
            // mismatched slot key would only fail server-side — a
            // mismatch reads as awaiting-the-key, i.e. the restore path.
            if (!keyOnDevice(status.actorPubkey)) return RegistrationProgress.AwaitingSigningKey
            return when (val result = writeSigner.signStaged(staged)) {
                is WriteResult.Done, is WriteResult.AwaitingSeal -> RegistrationProgress.AwaitingLanding
                is WriteResult.Refused -> RegistrationProgress.Refused(result.errors)
                is WriteResult.RejectedByDevice -> RegistrationProgress.RejectedByDevice(result.reason)
                is WriteResult.Failed -> RegistrationProgress.Failed(result.cause)
            }
        }
        val application = status.application ?: return RegistrationProgress.NeedsInvite
        if (application.approvedAt != null) return RegistrationProgress.AwaitingLanding
        if (application.expiresAt.isBefore(Instant.now())) return RegistrationProgress.NeedsInvite
        return RegistrationProgress.AwaitingApproval(
            emailVerified = application.emailVerified,
            keyAttached = application.keyAttached || repairAttach(),
            keyOnDevice = keyOnDevice(status.actorPubkey),
        )
    }

    /**
     * Whether the key in this account's slot is the one the account
     * runs on: a key is held AND (none is attached yet OR the slot key
     * IS the attached one). A mismatch reads as key-not-on-device, so
     * the UI offers the restore path (auth.md "Multi-account device
     * custody").
     */
    private suspend fun keyOnDevice(attachedPubkey: String?): Boolean {
        val slot = ceremony.publicIdentity() ?: return false
        return attachedPubkey == null || attachedPubkey == slot.publicKeyBase64
    }

    /**
     * Crash healing for the gap between minting and attaching: a key
     * that exists only on this device is re-attached silently. It fires
     * only when the server reports no key attached (the call site's
     * gate) and this account's own slot holds one — per-account keying
     * is what keeps another account's key out of reach. False when
     * there is nothing to repair or the attach did not land.
     */
    private suspend fun repairAttach(): Boolean {
        if (identity.actorSeed() == null) return false
        return ceremony.attachActorKey() is Outcome.Success
    }
}

File: web/src/lib/signing/registration-signer.ts
// One pass of the application poll (auth.md "Approval and landing";
// Android mirror: RegistrationSigner.kt): read the status, flush the
// parked backup, then branch — member, staged Registration to sign,
// re-arm needed, landing awaited, or the applicant cards.

import type { ApolloClient } from "@apollo/client";

import type { UserError } from "@/lib/api/outcome";
import { fetchApplicationStatus } from "@/lib/api/onboarding-api";
import { toBase64 } from "@/lib/crypto/bytes";
import type { KeyCeremony } from "@/lib/identity/key-ceremony";
import type { IdentityStore } from "@/lib/identity/store";
import type { AuthGuard } from "@/lib/session/guard";
import { keyUsable } from "./key-usable";
import type { WriteSigner } from "./write-signer";

export type RegistrationProgress =
  /** Landed — the loop's terminal state. */
  | { kind: "member" }
  /** Approved and signed; the wait is on the server. */
  | { kind: "awaitingLanding" }
  /** A staged Registration exists but this device cannot sign — restore the key. */
  | { kind: "awaitingSigningKey" }
  /** The live application: the two proofs and whether the key is on this device. */
  | {
      kind: "awaitingApproval";
      emailVerified: boolean;
      keyAttached: boolean;
      keyOnDevice: boolean;
    }
  /** No application, or an expired one — a fresh invite re-arms it. */
  | { kind: "needsInvite" }
  /** The device refused to sign — no amount of polling repairs it. */
  | { kind: "rejectedByDevice"; reason: string }
  | { kind: "refused"; errors: readonly UserError[] }
  | { kind: "failed"; cause: unknown };

export type RegistrationSigner = {
  advance(): Promise<RegistrationProgress>;
};

export function createRegistrationSigner(deps: {
  client: ApolloClient;
  guard: AuthGuard;
  store: IdentityStore;
  ceremony: KeyCeremony;
  writeSigner: WriteSigner;
  /** Injectable for tests; production uses the wall clock. */
  now?: () => number;
}): RegistrationSigner {
  const { client, guard, store, ceremony, writeSigner } = deps;
  const now = deps.now ?? (() => Date.now());

  /** base64 of the account slot's public key half; null when the slot is empty. */
  async function devicePubkey(): Promise<string | null> {
    const key = await store.actorKey();
    return key === null ? null : toBase64(key.publicKeyBytes());
  }

  /**
   * Silent crash healing for the gap between minting and attaching:
   * reached only when the server reports no key attached AND this
   * account's slot holds a key — never a blind fire across accounts.
   */
  async function repairAttach(): Promise<boolean> {
    return (await ceremony.attachActorKey()).kind === "success";
  }

  return {
    async advance() {
      const status = await guard.run(() => fetchApplicationStatus(client));
      if (status.kind === "refused") return { kind: "refused", errors: status.errors };
      if (status.kind === "failed") return { kind: "failed", cause: status.cause };

      // A session exists from registration on, so a parked blob never
      // waits on a milestone — flush every pass, result ignored.
      await ceremony.uploadPendingBackup();

      const { accountState, actorPubkey, application, stagedRegistration } = status.value;
      if (accountState === "MEMBER") return { kind: "member" };

      if (stagedRegistration !== null) {
        if (!keyUsable(await devicePubkey(), actorPubkey)) {
          return { kind: "awaitingSigningKey" };
        }
        const result = await writeSigner.signStaged(stagedRegistration);
        switch (result.kind) {
          case "done":
          case "awaitingSeal":
            return { kind: "awaitingLanding" };
          case "refused":
            return { kind: "refused", errors: result.errors };
          case "rejectedByDevice":
            return { kind: "rejectedByDevice", reason: result.reason };
          case "failed":
            return { kind: "failed", cause: result.cause };
        }
      }

      if (application === null) return { kind: "needsInvite" };
      if (application.approvedAt !== null) return { kind: "awaitingLanding" };
      if (Date.parse(application.expiresAt) < now()) return { kind: "needsInvite" };

      const devicePub = await devicePubkey();
      return {
        kind: "awaitingApproval",
        emailVerified: application.emailVerified,
        keyAttached: application.keyAttached || (devicePub !== null && (await repairAttach())),
        keyOnDevice: keyUsable(devicePub, actorPubkey),
      };
    },
  };
}
